doc/chicken-profile.1
.\" dummy line
.TH CHICKEN-PROFILE 1 "19 Sep 2001"

.SH NAME

chicken-profile \- generate a report from profile-information 

.SH SYNOPSIS

.B chicken-profile
[
.I filename
|
.I option ...
]

.SH OPTIONS

.TP
.B \-sort\-by\-calls
Sort output by call frequency.

.TP
.B \-sort\-by\-time
Sort output by procedure execution time.

.TP
.B \-sort\-by\-avg
Sort output by average procedure execution time.

.TP
.B \-sort\-by\-name
Sort output alphabetically by procedure name.

.TP
.B \-no\-unused
Remove procedures that are never called.

.TP
.B \-help
Show usage information.


.SH DESCRIPTION

.I chicken\-profile reads in profiling information generated by CHICKEN Scheme
programs compiled with the 
.B \-profile
option, and generates a table listing function names and their execution times.


.SH BUGS
Submit bug reports by e-mail to
.I chicken-janitors@nongnu.org

.SH AUTHORS
The CHICKEN Team

.SH SEE ALSO
.BR chicken(1)
.BR csc(1)


NEWS
4.9.1

- Security fixes
  - CVE-2014-6310: Use POSIX poll() on Android platform to avoid
    potential select() buffer overrun.

- Core libraries
  - alist-ref from unit data-structures now gives an error when passed
    a non-list, for consistency with assv/assq/assoc.
  - Unit tcp now implicitly depends on ports instead of extras.
     This may break programs which don't use modules and forgot to
     require extras but use procedures from it.
  - SRFI-13: fix string-copy! in cases source and destination strings'
    memory areas overlap (#1135).
  - SRFI-1: Check argument types in lset and list= procedures (#1085).
  - Fixed another, similar bug in move-memory! for overlapping memory.
  - Fixed broken specialisation for move-memory! on pointer types.
  - Fixed broken specialisation for irregex-match-num-submatches.
  - Fixed bug in make-kmp-restart-vector from SRFI-13.
  - Removed deprecated implicit expansion of $VAR- and ~ in pathnames.
     The ~-expansion functionality is now available in the
     "pathname-expand" egg (#1001, #1079) (thanks to Florian Zumbiehl).
  - normalize-pathname has been simplified to avoid adding trailing
     slashes or dots (#1153, thanks to Michele La Monaca and Mario Goulart).

- Unit srfi-4:
   - write-u8vector has been made more efficient (thanks to Thomas Hintz).

- Unit lolevel:
  - Restore long-lost but still documented "vector-like?" procedure (#983)

- Unit "posix":
  - set-file-position! now allows negative positions for seek/cur (thanks
    to Seth Alves).
  - file-mkstemp now works correctly on Windows, it now returns valid
    file descriptors (#819, thanks to Michele La Monaca).

- Runtime system:
  - Removed several deprecated, undocumented parts of the C interface:
    C_zap_strings, C_stack_check, C_retrieve, C_retrieve_proc,
    C_retrieve_symbol_proc, C_i_foreign_number_vector_argumentp,
    C_display_flonum, C_enumerate_symbols
  - Removed several deprecated and undocumented internal procedures:
    ##sys#zap-strings, ##sys#round, ##sys#foreign-number-vector-argument,
    ##sys#check-port-mode, ##sys#check-port*

- Module system
  - Allow functor arguments to be optional, with default implementations.
  - Fixed a bug that prevented functors from being instantiated with
     built-in modules.
  - Fixed generation of import libraries for instantiated functors
     (#1149, thanks to Juergen Lorenz).

- Syntax expander
  - define-values, set!-values and letrec-values now support full lambda
    lists as binding forms
  - cond expands correctly when a constant is used in combination with =>
     (thanks to Michele La Monaca)

- C API
  - Removed deprecated C_get_argument[_2] and
    C_get_environment_variable[_2] functions.
  - C_mutate2 has been deprecated in favor of C_mutate

- Foreign function interface
  - The foreign type specifier "scheme-pointer" now accepts an optional
    C pointer type (thanks to Moritz Heidkamp and Kristian Lein-Mathisen).
  - Type hinting for foreign-primitives now allows returning multiple
    values when no return type has been specified.

- Build system
  - MANDIR was renamed to MAN1DIR and TOPMANDIR was renamed to MANDIR
     in order to comply with standard Makefile practice in UNIX.
  - INCDIR was renamed to CHICKENINCDIR, and now derives from
     INCLUDEDIR, which defaults to $(PREFIX)/include
  - BINDIR, LIBDIR, SHAREDIR, INCLUDEDIR, MANDIR, MAN1DIR, DOCDIR,
     CHICKENINCDIR and CHICKENLIBDIR will now also be taken from
     the environment, if present (like PLATFORM, DESTDIR and PREFIX).

- Tools
  - "csc"
    - On Cygwin, -static now works again (thanks to Michele La Monaca)
  - "chicken-install"
    - When using chicken-install -retrieve, and an error occurs during
       retrieval (or the egg doesn't exist), the egg's directory is
       now properly cleaned up (#1109, thanks to Alex Charlton)
  - "chicken"
    - The -r5rs-syntax option did nothing; this has been fixed.

4.9.0

- Security fixes
  - CVE-2014-3776: read-u8vector! no longer reads beyond its buffer when
    length is #f (thanks to Seth Alves).
  - CVE-2013-4385: read-string! no longer reads beyond its buffer when
    length is #f.
  - CVE-2013-1874: ./.csirc is no longer loaded from the current directory
    upon startup of csi, which could lead to untrusted code execution.
    (thanks to Florian Zumbiehl)
  - CVE-2013-2024: On *nix, the qs procedure now single-quotes everything
    instead of relying on a blacklist of shell characters to be escaped.
    On Windows, it properly duplicates double-quote characters.  (thanks
    to Florian Zumbiehl)
  - CVE-2013-2075: Use POSIX poll() in other places where select() was
    still being used.  (thanks to Florian Zumbiehl and Joerg Wittenberger)
  - CVE-2012-6122: Use POSIX poll() on systems where available.  This avoids a
    design flaw in select(); it supports no more than FD_SETSIZE descriptors.

- Core libraries
  - Fix subvector when the TO optional argument equals the given vector
    length (#1097)
  - Unit extras now implicitly depends on ports.  ports no longer
     implicitly depends on extras.  This may break programs which don't
     use modules and forgot to require ports but use procedures from it.
  - Support has been added for the space-safe R7RS macro "delay-force".
  - Export file-type from the posix unit (thanks to Alan Post).
  - SRFI-4 s8vectors now work correctly in compiled code on PowerPC and ARM.
  - thread-join! now works correctly even if the waiting thread was
     prematurely woken up by a signal.
  - unsetenv has been fixed on Windows.
  - The process procedure has been fixed on Windows.
  - Nonblocking behaviour on sockets has been fixed on Windows.
  - Possible race condition while handling TCP errors has been fixed.
  - The posix unit will no longer hang upon any error in Windows.
  - resize-vector no longer crashes when reducing the size of the vector.
  - Distinct types for boolean true and false have been added to the
    scrutinizer.

src/chicken-profile.scm
;;;; chicken-profile.scm - Formatted display of profile outputs - felix -*- Scheme -*-
;
; Copyright (c) 2008-2014, The CHICKEN Team
; Copyright (c) 2000-2007, Felix L. Winkelmann
; All rights reserved.
;
; Redistribution and use in source and binary forms, with or without modification, are permitted provided that the following
; conditions are met:
;
;   Redistributions of source code must retain the above copyright notice, this list of conditions and the following
;     disclaimer.
;   Redistributions in binary form must reproduce the above copyright notice, this list of conditions and the following
;     disclaimer in the documentation and/or other materials provided with the distribution.
;   Neither the name of the author nor the names of its contributors may be used to endorse or promote
;     products derived from this software without specific prior written permission.
;
; THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS" AND ANY EXPRESS
; OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY
; AND FITNESS FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDERS OR
; CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
; CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
; SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
; THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR
; OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
; POSSIBILITY OF SUCH DAMAGE.


(declare
  (block)
  (uses srfi-1
        data-structures
        posix
        utils))

(define symbol-table-size 3001)

(define sort-by #f)
(define file #f)
(define no-unused #f)
(define seconds-digits 3)
(define average-digits 3)
(define percent-digits 3)
(define top 0)

(define (print-usage)
  (display #<#EOF
Usage: chicken-profile [FILENAME | OPTION] ...

 -sort-by-calls            sort output by call frequency
 -sort-by-time             sort output by procedure execution time
 -sort-by-avg              sort output by average procedure execution time
 -sort-by-name             sort output alphabetically by procedure name
 -decimals DDD             set number of decimals for seconds, average and
                           percent columns (three digits, default: #{seconds-digits}#{average-digits}#{percent-digits})
 -no-unused                remove procedures that are never called
 -top N                    display only the top N entries
 -help                     show this text and exit
 -version                  show version and exit
 -release                  show release number and exit

 FILENAME defaults to the `PROFILE.<number>', selecting the one with
 the highest modification time, in case multiple profiles exist.

EOF
;|
)
 (exit 64) )

(define (run args)
  (let loop ([args args])
    (if (null? args)
        (begin
          (unless file
            (set! file
              (let ((fs (glob "PROFILE.*")))
                (if (null? fs)
                    (error "no PROFILEs found")
                    (first (sort fs
                                 (lambda (f1 f2)
                                   (> (file-modification-time f1)
                                      (file-modification-time f2))) ) ) ) ) ) )
          (write-profile) )
        (let ([arg (car args)]
              [rest (cdr args)] )
          (define (next-arg)
            (if (null? rest)
                (error "missing argument to option" arg)
                (let ((narg (car rest)))
                  (set! rest (cdr rest))
                  narg)))
          (define (next-number)
            (let ((n (string->number (next-arg))))
              (if (and n (> n 0)) n (error "invalid argument to option" arg))))
          (cond
           [(member arg '("-h" "-help" "--help")) (print-usage)]
           [(string=? arg "-version")
            (print "chicken-profile - Version " (chicken-version))
            (exit) ]
           [(string=? arg "-release")
            (print (chicken-version))
            (exit) ]
           [(string=? arg "-no-unused") (set! no-unused #t)]
           [(string=? arg "-top") (set! top (next-number))]
           [(string=? arg "-sort-by-calls") (set! sort-by sort-by-calls)]
           [(string=? arg "-sort-by-time") (set! sort-by sort-by-time)]
           [(string=? arg "-sort-by-avg") (set! sort-by sort-by-avg)]
           [(string=? arg "-sort-by-name") (set! sort-by sort-by-name)]
           [(string=? arg "-decimals") (set-decimals (next-arg))]
           [(and (> (string-length arg) 1) (char=? #\- (string-ref arg 0)))
            (error "invalid option" arg) ]
           [file (print-usage)]
           [else (set! file arg)] )
          (loop rest) ) ) ) )

(define (sort-by-calls x y)
  (let ([c1 (second x)]
        [c2 (second y)] )
    (if (eqv? c1 c2)
        (> (third x) (third y))
        (if c1 (if c2 (> c1 c2) #t) #t) ) ) )

(define (sort-by-time x y)
  (let ([c1 (third x)]
        [c2 (third y)] )
    (if (= c1 c2)
        (> (second x) (second y))
        (> c1 c2) ) ) )

(define (sort-by-avg x y)
  (let ([c1 (cadddr x)]
        [c2 (cadddr y)] )
    (if (eqv? c1 c2)
        (> (third x) (third y))
        (> c1 c2) ) ) )

(define (sort-by-name x y)
  (string<? (symbol->string (first x)) (symbol->string (first y))) )

(set! sort-by sort-by-time)

(define (set-decimals arg)
  (if (= (string-length arg) 3)
      (begin
        (define (arg-digit n)
          (let ((n (- (char->integer (string-ref arg n))
                      (char->integer #\0))))
            (if (<= 0 n 9)
                (if (= n 9) 8 n) ; 9 => overflow in format-real
                (error "invalid argument to -decimals option" arg))))
        (set! seconds-digits (arg-digit 0))
        (set! average-digits (arg-digit 1))
        (set! percent-digits (arg-digit 2)))
      (error "invalid argument to -decimals option" arg)))

(define (make-sysmbol-table)
  (make-vector symbol-table-size '()))

(define (read-profile)
  (let ((hash (make-symbol-table)))
    (do ((line (read) (read)))
        ((eof-object? line))
      (##sys#hash-table-set!
       hash (first line)
       (map (lambda (x y) (and x y (+ x y)))
            (or (##sys#hash-table-ref hash (first line)) '(0 0))
            (cdr line))))
    (let ((alist '()))
      (##sys#hash-table-for-each
       (lambda (sym counts)
         (set! alist (alist-cons sym counts alist)))
       hash)
      alist)))

(define (format-string str cols #!optional right (padc #\space))
  (let* ((len (string-length str))
         (pad (make-string (fxmax 0 (fx- cols len)) padc)) )
    (if right
        (string-append pad str)
        (string-append str pad) ) ) )

(define (format-real n d)
  (let ((exact-value (inexact->exact (truncate n))))
    (string-append
     (number->string exact-value)
     (if (> d 0) "." "")
     (substring
      (number->string
       (inexact->exact
        (truncate
         (* (- n exact-value -1) (expt 10 d)))))
      1 (+ d 1)))))

(define (write-profile)
  (print "reading `" file "' ...\n")
  (let* ((data0 (with-input-from-file file read-profile))
         (max-t (foldl (lambda (r t) (max r (third t))) 0 data0))
         (data (sort (map
                      (lambda (t)
                        (append
                         t
                         (let ((c (second t)) ; count
                               (t (third t))) ; total time
                           (list (or (and c (> c 0) (/ t c)) ; time / count
                                     0)
                                 (or (and (> max-t 0) (* (/ t max-t) 100)) ; % of max-time
                                     0)
                                 ))))
                      data0)
                     sort-by)))
    (if (< 0 top (length data))
        (set! data (take data top)))
    (set! data (map (lambda (entry)
                      (let ([c (second entry)] ; count
                            [t (third entry)]  ; total time
                            [a (fourth entry)] ; average time
                            [p (fifth entry)] ) ; % of max time
                        (list (##sys#symbol->qualified-string (first entry))
                              (if (not c) "overflow" (number->string c))
                              (format-real (/ t 1000) seconds-digits)
                              (format-real (/ a 1000) average-digits)
                              (format-real p percent-digits))))
                    (remove (lambda (entry)
                              (if (second entry)
                                  (and (zero? (second entry)) no-unused)
                                  #f) )
                            data)))
    (let* ([headers (list "procedure" "calls" "seconds" "average" "percent")]
           [alignments (list #f #t #t #t #t)]
           [spacing 2]
           [spacer (make-string spacing #\space)]
           [column-widths (fold
                           (lambda (row max-widths)
                             (map max (map string-length row) max-widths))
                           (list 0 0 0 0 0)
                           (cons headers data))])
      (define (print-row row)
        (print (string-intersperse (map format-string row column-widths alignments) spacer)))
      (print-row headers)
      (print (make-string (+ (reduce + 0 column-widths)
                             (* spacing (- (length alignments) 1)))
                          #\-))
      (for-each print-row data))))

(run (command-line-arguments))
